File: app/src/main/java/com/example/inventoryobat/TambahObatActivity.java
package com.example.inventoryobat;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.bumptech.glide.Glide;
import com.example.inventoryobat.databinding.ActivityTambahObatBinding;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class TambahObatActivity extends AppCompatActivity {

    private ActivityTambahObatBinding binding;
    private MainViewModel viewModel;
    private ApiService apiService;
    private Uri selectedImageUri;
    private List<Supplier> supplierList = new ArrayList<>();
    private int selectedSupplierId;
    private boolean isEdit = false;
    private int editObatId = -1;
    private static final int PICK_IMAGE_REQUEST = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityTambahObatBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        viewModel = new MainViewModel();
        apiService = ApiConfig.getApiService();

        isEdit = getIntent().getBooleanExtra("edit_mode", false);
        if (isEdit) {
            editObatId = getIntent().getIntExtra("obat_id", -1);
            binding.btnSimpan.setText("UPDATE OBAT");
        }

        if (!PermissionHelper.checkPermission(this)) {
            PermissionHelper.requestPermission(this);
        }

        setupJenisObatSpinner();
        setupSupplierSpinner();

        if (isEdit) {
            new Handler().postDelayed(this::loadDataForEdit, 200);
        }

        binding.btnPilihGambar.setOnClickListener(v -> openImagePicker());
        binding.btnSimpan.setOnClickListener(v -> simpanObat());
    }

    private void setupJenisObatSpinner() {
        ArrayList<String> jenisObatList = new ArrayList<>();
        for (JenisObat jenis : JenisObat.values()) {
            jenisObatList.add(jenis.getDisplayName());
        }

        ArrayAdapter<String> adapter = new ArrayAdapter<>(
                this,
                android.R.layout.simple_spinner_item,
                jenisObatList
        );
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        binding.spinnerJenisObat.setAdapter(adapter);
    }

    private void setupSupplierSpinner() {
        viewModel.loadAllSuppliers();
        viewModel.getSupplierList().observe(this, suppliers -> {
            supplierList = suppliers;
            ArrayList<String> names = new ArrayList<>();
            for (Supplier s : suppliers) {
                names.add(s.getNamaSupplier());
            }

            ArrayAdapter<String> adapter = new ArrayAdapter<>(
                    this,
                    android.R.layout.simple_spinner_item,
                    names
            );
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            binding.spinnerSupplier.setAdapter(adapter);

            binding.spinnerSupplier.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override
                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                    if (position >= 0 && position < supplierList.size()) {
                        selectedSupplierId = supplierList.get(position).getIdSupplier();
                    }
                }

                @Override
                public void onNothingSelected(AdapterView<?> parent) {}
            });
        });
    }

    private void loadDataForEdit() {
        if (editObatId < 0) return;

        viewModel.getObatById(editObatId).observe(this, obat -> {
            if (obat == null) return;

            binding.edtNamaObat.setText(obat.getNamaObat());
            binding.edtStock.setText(String.valueOf(obat.getStock()));

            String jenisObat = obat.getJenisObat();
            ArrayAdapter<String> adapter = (ArrayAdapter<String>) binding.spinnerJenisObat.getAdapter();
            if (adapter != null) {
                int position = adapter.getPosition(jenisObat);
                binding.spinnerJenisObat.setSelection(position >= 0 ? position : 0);
            }

            for (int i = 0; i < supplierList.size(); i++) {
                if (supplierList.get(i).getIdSupplier() == obat.getIdSupplier()) {
                    binding.spinnerSupplier.setSelection(i);
                    selectedSupplierId = obat.getIdSupplier();
                    break;
                }
            }

            if (obat.getGambarUrl() != null && !obat.getGambarUrl().isEmpty()) {
                Glide.with(this)
                        .load(obat.getGambarUrl())
                        .placeholder(R.drawable.ic_launcher_foreground)
                        .error(R.drawable.ic_launcher_foreground)
                        .into(binding.imgPreview);
            }
        });
    }

    private void openImagePicker() {
        if (!PermissionHelper.checkPermission(this)) {
            PermissionHelper.requestPermission(this);
            Toast.makeText(this, "Izin akses media diperlukan", Toast.LENGTH_SHORT).show();
            return;
        }

        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        intent.setType("image/*");
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION
                | Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION);
        startActivityForResult(intent, PICK_IMAGE_REQUEST);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PermissionHelper.PERMISSION_REQUEST_CODE &&
                grantResults.length > 0 &&
                grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            Toast.makeText(this, "Izin diberikan", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == PICK_IMAGE_REQUEST && resultCode == RESULT_OK && data != null) {
            Uri uri = data.getData();
            if (uri != null) {
                selectedImageUri = uri;
                try {
                    getContentResolver().takePersistableUriPermission(uri, Intent.FLAG_GRANT_READ_URI_PERMISSION);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                binding.imgPreview.setImageURI(uri);
            }
        }
    }

    private void simpanObat() {
        String nama = binding.edtNamaObat.getText().toString().trim();
        String stockStr = binding.edtStock.getText().toString().trim();

        if (nama.isEmpty() || stockStr.isEmpty()) {
            Toast.makeText(this, "Isi semua field!", Toast.LENGTH_SHORT).show();
            return;
        }

        int stock = Integer.parseInt(stockStr);
        String jenis = (String) binding.spinnerJenisObat.getSelectedItem();

        RequestBody namaBody = RequestBody.create(nama, MediaType.parse("text/plain"));
        RequestBody jenisBody = RequestBody.create(jenis, MediaType.parse("text/plain"));
        RequestBody stockBody = RequestBody.create(String.valueOf(stock), MediaType.parse("text/plain"));
        RequestBody supplierBody = RequestBody.create(String.valueOf(selectedSupplierId), MediaType.parse("text/plain"));

        MultipartBody.Part gambarPart = null;
        if (selectedImageUri != null) {
            File file = uriToFile(selectedImageUri);
            if (file != null) {
                RequestBody requestBody = RequestBody.create(file, MediaType.parse("image/*"));
                gambarPart = MultipartBody.Part.createFormData("gambar", file.getName(), requestBody);
            }
        }

        if (isEdit) {
            apiService.updateObat(editObatId, namaBody, jenisBody, stockBody, supplierBody, gambarPart)
                .enqueue(new Callback<ApiResponse<Obat>>() {
                    @Override
                    public void onResponse(Call<ApiResponse<Obat>> call, Response<ApiResponse<Obat>> response) {
                        if (response.isSuccessful()) {
                            Toast.makeText(TambahObatActivity.this, "Obat berhasil diupdate!", Toast.LENGTH_SHORT).show();
                            finish();
                        } else {
                            Toast.makeText(TambahObatActivity.this, "Gagal update obat: " + response.code(), Toast.LENGTH_SHORT).show();
                        }
                    }

                    @Override
                    public void onFailure(Call<ApiResponse<Obat>> call, Throwable t) {
                        Toast.makeText(TambahObatActivity.this, "Error: " + t.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
        } else {
            apiService.createObat(namaBody, jenisBody, stockBody, supplierBody, gambarPart)
                .enqueue(new Callback<ApiResponse<Obat>>() {
                    @Override
                    public void onResponse(Call<ApiResponse<Obat>> call, Response<ApiResponse<Obat>> response) {
                        if (response.isSuccessful()) {
                            Toast.makeText(TambahObatActivity.this, "Obat berhasil ditambahkan!", Toast.LENGTH_SHORT).show();
                            finish();
                        } else {
                            Toast.makeText(TambahObatActivity.this, "Gagal tambah obat: " + response.code(), Toast.LENGTH_SHORT).show();
                        }
                    }

                    @Override
                    public void onFailure(Call<ApiResponse<Obat>> call, Throwable t) {
                        Toast.makeText(TambahObatActivity.this, "Error: " + t.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
        }
    }

    private File uriToFile(Uri uri) {
        try {
            InputStream inputStream = getContentResolver().openInputStream(uri);
            File file = new File(getCacheDir(), "temp_image.jpg");
            FileOutputStream outputStream = new FileOutputStream(file);

            byte[] buffer = new byte[1024];
            int length;
            while ((length = inputStream.read(buffer)) > 0) {
                outputStream.write(buffer, 0, length);
            }

            outputStream.close();
            inputStream.close();
            return file;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: app/src/main/java/com/example/inventoryobat/TambahObatActivity.kt
package com.example.inventoryobat

import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.example.inventoryobat.databinding.ActivityTambahObatBinding
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.File
import java.io.FileOutputStream

class TambahObatActivity : AppCompatActivity() {

    private lateinit var binding: ActivityTambahObatBinding
    private lateinit var viewModel: MainViewModel
    private lateinit var apiService: ApiService
    private var selectedImageUri: Uri? = null
    private var supplierList: List<Supplier> = emptyList()
    private var selectedSupplierId = 0
    private var isEdit = false
    private var editObatId = -1

    companion object {
        private const val PICK_IMAGE_REQUEST = 1
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityTambahObatBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = MainViewModel()
        apiService = ApiConfig.getApiService()

        isEdit = intent.getBooleanExtra("edit_mode", false)
        if (isEdit) {
            editObatId = intent.getIntExtra("obat_id", -1)
            binding.btnSimpan.text = "UPDATE OBAT"
        }

        if (!PermissionHelper.checkPermission(this)) {
            PermissionHelper.requestPermission(this)
        }

        setupJenisObatSpinner()
        setupSupplierSpinner()

        if (isEdit) {
            Handler(Looper.getMainLooper()).postDelayed({ loadDataForEdit() }, 200)
        }

        binding.btnPilihGambar.setOnClickListener { openImagePicker() }
        binding.btnSimpan.setOnClickListener { simpanObat() }
    }

    private fun setupJenisObatSpinner() {
        val jenisObatList = JenisObat.entries.map { it.displayName }
        val adapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, jenisObatList)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.spinnerJenisObat.adapter = adapter
    }

    private fun setupSupplierSpinner() {
        viewModel.loadAllSuppliers()
        viewModel.supplierList.observe(this) { suppliers ->
            supplierList = suppliers
            val names = suppliers.map { it.namaSupplier }
            val adapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, names)
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            binding.spinnerSupplier.adapter = adapter

            binding.spinnerSupplier.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                    if (position in supplierList.indices) {
                        selectedSupplierId = supplierList[position].idSupplier
                    }
                }
                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        }
    }

    private fun loadDataForEdit() {
        if (editObatId < 0) return

        viewModel.getObatById(editObatId).observe(this) { obat ->
            obat ?: return@observe

            binding.edtNamaObat.setText(obat.namaObat)
            binding.edtStock.setText(obat.stock.toString())

            val spinnerAdapter = binding.spinnerJenisObat.adapter as? ArrayAdapter<String>
            spinnerAdapter?.let {
                val position = it.getPosition(obat.jenisObat)
                binding.spinnerJenisObat.setSelection(if (position >= 0) position else 0)
            }

            for (i in supplierList.indices) {
                if (supplierList[i].idSupplier == obat.idSupplier) {
                    binding.spinnerSupplier.setSelection(i)
                    selectedSupplierId = obat.idSupplier
                    break
                }
            }

            if (obat.gambarUrl.isNotEmpty()) {
                Glide.with(this)
                    .load(obat.gambarUrl)
                    .placeholder(R.drawable.ic_launcher_foreground)
                    .error(R.drawable.ic_launcher_foreground)
                    .into(binding.imgPreview)
            }
        }
    }

    private fun openImagePicker() {
        if (!PermissionHelper.checkPermission(this)) {
            PermissionHelper.requestPermission(this)
            Toast.makeText(this, "Izin akses media diperlukan", Toast.LENGTH_SHORT).show()
            return
        }

        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            type = "image/*"
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION)
        }
        startActivityForResult(intent, PICK_IMAGE_REQUEST)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PermissionHelper.PERMISSION_REQUEST_CODE &&
            grantResults.isNotEmpty() &&
            grantResults[0] == PackageManager.PERMISSION_GRANTED
        ) {
            Toast.makeText(this, "Izin diberikan", Toast.LENGTH_SHORT).show()
        }
    }

    @Deprecated("Use Activity Result API")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == PICK_IMAGE_REQUEST && resultCode == RESULT_OK && data != null) {
            data.data?.let { uri ->
                selectedImageUri = uri
                try {
                    contentResolver.takePersistableUriPermission(uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
                } catch (_: Exception) { }
                binding.imgPreview.setImageURI(uri)
            }
        }
    }

    private fun simpanObat() {
        val nama = binding.edtNamaObat.text.toString().trim()
        val stockStr = binding.edtStock.text.toString().trim()

        if (nama.isEmpty() || stockStr.isEmpty()) {
            Toast.makeText(this, "Isi semua field!", Toast.LENGTH_SHORT).show()
            return
        }

        val stock = stockStr.toInt()
        val jenis = binding.spinnerJenisObat.selectedItem as String
        val textPlain = "text/plain".toMediaTypeOrNull()

        if (isEdit) {
            val uri = selectedImageUri
            if (uri != null) {
                editWithImage(nama, jenis, stock, uri, textPlain)
            } else {
                editWithoutImage(nama, jenis, stock)
            }
        } else {
            createObat(nama, jenis, stock, textPlain)
        }
    }

    private fun editWithImage(nama: String, jenis: String, stock: Int, uri: Uri, textPlain: okhttp3.MediaType?) {
        val file = uriToFile(uri) ?: return
        val requestBody = file.asRequestBody("image/*".toMediaTypeOrNull())
        val gambarPart = MultipartBody.Part.createFormData("gambar", file.name, requestBody)

        apiService.updateObatWithImage(
            editObatId,
            nama.toRequestBody(textPlain),
            jenis.toRequestBody(textPlain),
            stock.toString().toRequestBody(textPlain),
            selectedSupplierId.toString().toRequestBody(textPlain),
            gambarPart,
            "PUT".toRequestBody(textPlain)
        ).enqueue(object : Callback<ApiResponse<Obat>> {
            override fun onResponse(call: Call<ApiResponse<Obat>>, response: Response<ApiResponse<Obat>>) {
                if (response.isSuccessful) {
                    Toast.makeText(this@TambahObatActivity, "Update (Img) Berhasil!", Toast.LENGTH_SHORT).show()
                    finish()
                } else {
                    Toast.makeText(this@TambahObatActivity, "Gagal Update: ${response.code()}", Toast.LENGTH_SHORT).show()
                }
            }
            override fun onFailure(call: Call<ApiResponse<Obat>>, t: Throwable) {
                Toast.makeText(this@TambahObatActivity, "Error: ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun editWithoutImage(nama: String, jenis: String, stock: Int) {
        apiService.updateObatNoImage(editObatId, nama, jenis, stock, selectedSupplierId)
            .enqueue(object : Callback<ApiResponse<Obat>> {
                override fun onResponse(call: Call<ApiResponse<Obat>>, response: Response<ApiResponse<Obat>>) {
                    if (response.isSuccessful) {
                        Toast.makeText(this@TambahObatActivity, "Update Data Berhasil!", Toast.LENGTH_SHORT).show()
                        finish()
                    } else {
                        Toast.makeText(this@TambahObatActivity, "Gagal Update: ${response.code()}", Toast.LENGTH_SHORT).show()
                    }
                }
                override fun onFailure(call: Call<ApiResponse<Obat>>, t: Throwable) {
                    Toast.makeText(this@TambahObatActivity, "Error: ${t.message}", Toast.LENGTH_SHORT).show()
                }
            })
    }

    private fun createObat(nama: String, jenis: String, stock: Int, textPlain: okhttp3.MediaType?) {
        val namaBody = nama.toRequestBody(textPlain)
        val jenisBody = jenis.toRequestBody(textPlain)
        val stockBody = stock.toString().toRequestBody(textPlain)
        val supplierBody = selectedSupplierId.toString().toRequestBody(textPlain)

        var gambarPart: MultipartBody.Part? = null
        selectedImageUri?.let { uri ->
            uriToFile(uri)?.let { file ->
                val requestBody = file.asRequestBody("image/*".toMediaTypeOrNull())
                gambarPart = MultipartBody.Part.createFormData("gambar", file.name, requestBody)
            }
        }

        val part = gambarPart
        if (part == null) {
            Toast.makeText(this, "Pilih gambar terlebih dahulu", Toast.LENGTH_SHORT).show()
            return
        }

        apiService.createObat(namaBody, jenisBody, stockBody, supplierBody, part)
            .enqueue(object : Callback<ApiResponse<Obat>> {
                override fun onResponse(call: Call<ApiResponse<Obat>>, response: Response<ApiResponse<Obat>>) {
                    if (response.isSuccessful) {
                        Toast.makeText(this@TambahObatActivity, "Obat berhasil ditambahkan!", Toast.LENGTH_SHORT).show()
                        finish()
                    } else {
                        Toast.makeText(this@TambahObatActivity, "Gagal tambah obat: ${response.code()}", Toast.LENGTH_SHORT).show()
                    }
                }
                override fun onFailure(call: Call<ApiResponse<Obat>>, t: Throwable) {
                    Toast.makeText(this@TambahObatActivity, "Error: ${t.message}", Toast.LENGTH_SHORT).show()
                }
            })
    }

    private fun uriToFile(uri: Uri): File? {
        return try {
            val inputStream = contentResolver.openInputStream(uri) ?: return null
            val file = File(cacheDir, "temp_image_${System.currentTimeMillis()}.jpg")
            FileOutputStream(file).use { outputStream ->
                inputStream.use { input -> input.copyTo(outputStream) }
            }
            file
        } catch (_: Exception) {
            null
        }
    }
}
